### PR TITLE
fix(sight): match tool args in loop detection

### LINKED ISSUE #2691: [P2][sight] tool_sequence_repetition flags consecutive same-tool calls as dead_loop (critical)
### Component

sight

### What happened?

Agents that were working normally (multi-step tasks, all completed successfully) were flagged with `dead_loop` / `critical` interruption events by the `tool_sequence_repetition` rule:

```json
- {"repeated_tools":["terminal"],"repeat_count":5,"rule":"tool_sequence_repetition"}
- {"repeated_tools":["exec"],"repeat_count":5,"rule":"tool_sequence_repetition"}
```

In one flagged case the 5 consecutive `terminal` calls each ran a *different* command (`pip list` → `find` → `cat` → a detect script → an install script), made clear progress, and the task finished successfully — there was no loop.

What I expected: normal multi-step work that naturally invokes the same tool repeatedly (terminal/exec-heavy workflows) should not be classified as a critical dead loop.

### Root cause analysis

In `src/agentsight/src/interruption/loop_detector.rs`, rule 1 (`detect_tool_sequence_loop`) compares only the **tool name sequence** (`tool_call_names`) of the last N tool-bearing calls. It does not look at the tool call arguments/commands, so five different commands executed via the same tool are indistinguishable from a genuine loop.

Additionally:

- The threshold `tool_sequence_repeat_threshold` exists in `LoopDetectorConfig`, but the call site in `src/agentsight/src/unified.rs` constructs `LoopDetector::default()`, so the config value is effectively hardcoded to 5 and cannot be tuned via config.json.
- `InterruptionType::DeadLoop` has a fixed default severity of `Critical` (`src/agentsight/src/interruption/types.rs`), so every false positive is reported at the highest severity.

### How can we reproduce it?

1. Run `agentsight trace` with interruption detection enabled (default).
2. Have any agent perform a normal multi-step task that issues 5+ consecutive calls to the same tool with different arguments (e.g. a shell/terminal tool running different commands).
3. Observe a `dead_loop` / `critical` event with `rule: tool_sequence_repetition` even though the task progresses and completes.

### Anything else?

Suggested fixes:

1. Improve precision: require the *same tool arguments/command* to repeat (not just the same tool name) before flagging a dead loop. This needs `RecentCallSummary` to carry an argument fingerprint.
2. Wire `tool_sequence_repeat_threshold` (and related loop-detector knobs) into config.json instead of always using `LoopDetector::default()` at the call site.
3. Reconsider severity: name-only repetition could be reported at a lower severity, reserving `critical` for repetitions with identical arguments.

### PR DESCRIPTION
**Pull request by jfeng18**
## Why

`tool_sequence_repetition` compares only the tool-name sequence of recent calls, so agents doing normal multi-step work (e.g. five consecutive `terminal` calls each running a different command) are flagged as `dead_loop` / `critical`. See #2691.

## What changed

- Add `ToolCallKey { name, args_fingerprint }` to the loop detector: Rule 1 now counts a repetition only when both the tool name and the argument fingerprint match.
- Fingerprints are computed in `parse_output_messages_for_loop_detection` by hashing the serialized `arguments` of each `tool_call` part.
- Calls without arguments keep the legacy name-only comparison (`None == None`), so providers that never emit arguments retain the existing detection capability.

Out of scope (kept as follow-ups per issue suggestions 2/3): wiring `tool_sequence_repeat_threshold` into config.json (behavior-neutral today since the call site uses defaults), and severity re-grading (obsolete once name-only repetition no longer fires).

## Related issue

closes #2691

## User / Agent impact

Fewer false `dead_loop`/`critical` interruption events for terminal/exec-heavy workflows; genuine loops (same tool, same arguments) are still detected. With `deadloop.kill` enabled, this also removes a spurious auto-kill trigger.

## Risk and compatibility

- [ ] Public CLI, API, configuration, or documented behavior changed
- [ ] Privileged or security-sensitive behavior changed
- [ ] Cross-component contract changed
- [ ] Migration or rollback guidance is needed

Low risk. Detection-only change; the `detail` JSON shape (`repeated_tools`, `repeat_count`, `rule`) is unchanged. serde_json runs with `preserve_order`, so same-source repeated calls serialize identically; order-differing but semantically equal arguments hash differently, which only makes detection more conservative.

## Validation

On ECS (Alinux, kernel 6.6, rustup 1.89.0 matching CI):

- `cargo fmt --all --check` — pass
- `python3 scripts/check-arch-boundaries.py` — 0 violations
- `cargo clippy --workspace --all-targets -- -D warnings` — clean
- `cargo test --workspace` — 1865 passed; the single failure is a pre-existing doctest in `src/genai/helpers.rs` (markdown in a doc comment compiled as a doctest, introduced by 93ccdd06b; CI's llvm-cov does not run doctests) — reported separately.
- New discriminating tests: `test_same_tool_different_args_not_a_loop` (fails if the fix is reverted), `test_same_tool_same_args_loop_detected` (guards against detection regression), `test_args_presence_mismatch_not_a_loop`, `test_fingerprint_args_stability`, `test_parse_output_tool_call_args_fingerprint`.

## Documentation and rollback

Doc comments updated in-module; no user-facing docs affected. Revert the single commit to roll back.

### CODE REVIEW

**Review by chengshuyi — APPROVED** [review]
LGTM. Clean, well-scoped fix. The (name, args_fingerprint) comparison correctly eliminates false positives for same-tool-different-args workflows while preserving genuine loop detection. Output JSON shape unchanged, backward compatible for no-args providers via None==None. Tests are thorough including the #2691 regression case. Minor nit: consider adding a test for explicit "arguments": null (Value::Null vs None edge), non-blocking.

File: src/agentsight/src/storage/sqlite/genai/pending.rs [hunk 1]
```diff
@@ -774,7 +777,13 @@ pub(super) fn parse_output_messages_for_loop_detection(
                 match part.get("type").and_then(|t| t.as_str()) {
                     Some("tool_call") => {
                         if let Some(name) = part.get("name").and_then(|n| n.as_str()) {
-                            tool_names.push(name.to_string());
+                            let args_fingerprint = part
+                                .get("arguments")
+                                .map(crate::interruption::ToolCallKey::fingerprint_args);
```
[the remark this answers is not in the document]
Comment (jfeng18) [reply src/agentsight/src/storage/sqlite/genai/pending.rs:782]: Fixed: added `loop_detection_reads_tool_args_through_store_roundtrip`, which persists real `MessagePart::ToolCall { arguments }` events, reads them back via `get_recent_calls_for_conversation`, and runs `LoopDetector` end to end — asserting that differing arguments do not trigger while identical ones do, and that fingerprints survive the storage round-trip (so a stored-shape drift fails this test). Insertion uses `try_insert_event` (same serialization path as `store_event`) to stay immune to the process-global size-limit env var that parallel size-limit tests set to 1MB.

File: src/agentsight/src/interruption/loop_detector.rs [hunk 2]
```diff
@@ -38,12 +39,42 @@ impl Default for LoopDetectorConfig {
...
+pub struct ToolCallKey {
+    pub name: String,
+    /// Stable hash of the serialized JSON arguments; `None` when the provider
+    /// omitted arguments. `None == None`, which keeps the legacy name-only
+    /// comparison for providers that never emit arguments.
+    pub args_fingerprint: Option<u64>,
+}
+
+impl ToolCallKey {
+    /// Hash the serialized arguments into a stable per-process fingerprint.
```
[the remark this answers is not in the document]
Comment (jfeng18) [reply src/agentsight/src/interruption/loop_detector.rs:57]: Not a real issue: the fingerprint is never persisted or exported — `get_recent_calls_for_conversation` recomputes it from the stored `output_messages` JSON on every detection pass, and the comparison happens within that single pass. Cross-process/version stability therefore has no correctness impact (a hash change alters nothing observable). The doc comment on `fingerprint_args` states this per-process scope.

File: src/agentsight/src/interruption/loop_detector.rs [hunk 3]
```diff
@@ -152,13 +182,14 @@ impl LoopDetector {
...
         let tail = &tool_bearing[tool_bearing.len().saturating_sub(threshold)..];
 
-        // Check if all tool sequences in the tail are identical
-        let reference = &tail[0].tool_call_names;
-        let all_same = tail[1..].iter().all(|c| &c.tool_call_names == reference);
+        // Check if all tool call sequences in the tail are identical
+        let reference = &tail[0].tool_calls;
+        let all_same = tail[1..].iter().all(|c| &c.tool_calls == reference);
 
         if all_same {
```
[the remark this answers is not in the document]
Comment (jfeng18) [reply src/agentsight/src/interruption/loop_detector.rs:189]: Intentional: `repeated_tools` keeps the existing name-array shape so dashboard/logtail consumers are unaffected, and raw arguments are deliberately not embedded in interruption details — these events are exported via logtail/SLS where tool arguments may carry sensitive payloads. The `rule` field (`tool_sequence_repetition`) already signals that detection now requires identical name+arguments.

File: src/agentsight/src/storage/sqlite/genai/pending.rs [hunk 4]
```diff
@@ -765,7 +768,7 @@ pub(super) fn parse_output_messages_for_loop_detection(
         Err(_) => return (vec![], String::new()),
```
[the remark this answers is not in the document]
Comment (jfeng18) [reply src/agentsight/src/storage/sqlite/genai/pending.rs:768]: Pre-existing behavior, not introduced by this diff: the silent early-return on parse error existed before; this PR only changes the returned type. This function parses JSON written by our own `store_event` in the same binary, so a failure here means corrupted rows. A debug log is a reasonable separate observability improvement; keeping this PR scoped.
